typing/typedecl_separability.ml

```diff
@@ -0,0 +1,674 @@
+        | Cstr_record labels ->
+            demultiply_list labels @@ fun ld ->
+              let argument_type = ld.ld_type in
+              {
+                location = ld.ld_loc;
+                kind = Constructor_field;
+                mutability = ld.ld_mutable;
+                argument_type;
+                result_type_parameter_instances;
+              }
+        end)
+
+
+type error =
+  | Non_separable_evar of string option
+
+exception Error of Location.t * error
+(*
+exception Not_separable of { loc: Location.t; evar: string option; }
+*)
+
+(* see the .mli file for explanations on the modes *)
+module Sep = Types.Separability
+type mode = Sep.t = Ind | Sep | Deepsep
+
+let rank = Sep.rank
+let max_mode = Sep.max
+
+(** If the type context [e(_)] imposes the mode [m] on its hole [_],
+    and the type context [e'(_)] imposes the mode [m'] on its hole [_],
+    then the mode on [_] imposed by the context composition [e(e'(_))]
+    is [compose m m'].
+
+    This operation differs from [max_mode]: [max_mode Ind Sep] is [Sep],
+    but [compose Ind Sep] is [Ind]. *)
+let compose
+  : mode -> mode -> mode
+  = fun m1 m2 ->
+  match m1 with
+  | Deepsep -> Deepsep
+  | Sep -> m2
+  | Ind -> Ind
+
+type type_var = {
+  text: string option; (** the user name of the type variable, None for '_' *)
+  id: int; (** the identifier of the type node (type_expr.id) of the variable *)
+}
+
+module TVarMap = Map.Make(struct
+    type t = type_var
+    let compare v1 v2 = compare v1.id v2.id
+  end)
+type context = mode TVarMap.t
+let (++) = TVarMap.union (fun _ m1 m2 -> Some(max_mode m1 m2))
+let empty = TVarMap.empty
+
+
+(** [immediate_subtypes ty] returns the list of all the
+   immediate sub-type-expressions of [ty]. They represent the biggest
+   sub-components that may be extracted using a constraint. For
+   example, the immediate sub-type-expressions of [int * (bool * 'a)]
+   are [int] and [bool * 'a].
+
+   Smaller components are extracted recursively in [check_type]. *)
+let rec immediate_subtypes : type_expr -> type_expr list = fun ty ->
+  (* Note: Btype.fold_type_expr is not suitable here:
+     - it does not do the right thing on Tpoly, iterating on type
+       parameters as well as the subtype
+     - it performs a shallow traversal of object types,
+       while our implementation collects all method types *)
+  match (Ctype.repr ty).desc with
+  (* these are the important cases,
+     on which immediate_subtypes is called from check_type_expr *)
+  | Tarrow(_,ty1,ty2,_) ->
+      [ty1; ty2]
+  | Ttuple(tys)
+  | Tpackage(_,_,tys) ->
+      tys
+  | Tobject(row,class_ty) ->
+      let class_subtys =
+        match !class_ty with
+        | None        -> []
+        | Some(_,tys) -> tys
+      in
+      immediate_subtypes_object_row class_subtys row
+  | Tvariant(row) ->
+      immediate_subtypes_variant_row [] row
+
+  (* the cases below are not called from check_type_expr,
+     they are here for completeness *)
```

Comment:
(I'm not fully sure that the comment is correct.) When we split some logic from a specific site into a generic function, it often happens that the natural specification of the generic function covers more than the interface used at the callsite. Either we make the specific more complex by ruling certain cases out, or we write dead code. In OCaml the former is made by adding `assert false`. The difference will be noticeable when someone uses the generic function in other places, or changes the caller in a way that invalidates the restrictions there; if you wrote `assert false`, then you will now blow up in some cases, if you handled these cases you run the risk of bugs because those paths were not tested. I would prefer the second alternative, relying on my great reviewers to also double-check that those cases make sense.